### ISSUE TITLE
General layout issues

### ISSUE THREAD

**Issue by jackdem** [MessageKit/MessageKit#268]
## General Information

* MessageKit Version: latest

* iOS Version(s): 11

* Swift Version: 3

* Devices/Simulators: 8 sim

* Reproducible in ChatExample? (Yes/No):

##  What happened?
I have several layout issues with the library, for example the chat window is hidden behind the message typing box, the view doesnt scroll to the bottom on load, opening the full keyboard doesnt life the chat view and obscures it. 

<img width="402" alt="screen shot 2017-10-18 at 11 59 56 am" src="https://user-images.githubusercontent.com/31346145/31715517-25b99344-b3fc-11e7-9bd3-3b02cb3dfd1f.png">

<img width="387" alt="screen shot 2017-10-18 at 12 02 28 pm" src="https://user-images.githubusercontent.com/31346145/31715546-40f497a8-b3fc-11e7-8f3c-bbf49ec126eb.png">

<img width="402" alt="screen shot 2017-10-18 at 12 02 55 pm" src="https://user-images.githubusercontent.com/31346145/31715561-4b7b92b2-b3fc-11e7-8cc6-1c264ecb068f.png">

## What did you expect to happen?
adapts correctly

**Comment by SD10** [comment]
Hey @jackdem,
There is a bug with the top content inset on master. It should be fixed with the next relase.

As for scrolling/keyboard handling. You should look at the `scrollsToBottomOnFirstLoad` and `scrollsToBottomOnKeyboardDidBeginEditing` properties of `MessagesViewController`. If there are any other layout problems please be specific 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by jackdem** [comment]
do i need to do any additional setup for those calls as either of them in the viewDidLoad doesnt scroll down correctly

```
override func viewDidLoad() {
        super.viewDidLoad()
        self.title = self.conversation.firstname + " " + self.conversation.lastname
        messagesCollectionView.messagesDataSource = self
        messagesCollectionView.messagesLayoutDelegate = self
        messagesCollectionView.messagesDisplayDelegate = self
        messagesCollectionView.messageCellDelegate = self
        messageInputBar.delegate = self
        scrollsToBottomOnFirstLayout = true
        scrollsToBottomOnKeybordBeginsEditing = true
        scheduledTimerWithTimeInterval()
        self.tabBarController?.tabBar.isHidden = true
    }
```

**Comment by SD10** [comment]
what do you mean by not scrolling down correctly @jackdem?

**Comment by jackdem** [comment]
as shown in the screenshots, when the view loads, itll load in the middle of a convo, it doesnt scroll to the latest message

**Comment by SD10** [comment]
Ok @jackdem I'll take a look, thanks for reporting this

**Comment by jackdem** [comment]
thanks, i assumed those library methods would allow the auto scroll as you mentioned, but as you can see in the viewdidload code above, im using them and its not triggering

**Comment by Terens777** [comment]
how did you solve this problem?

**Comment by jackdem** [comment]
i didnt, its messed up until the next release

**Comment by Terens777** [comment]
understood, we will work with it ourselves

**Comment by Terens777** [comment]
Yes, I did this, I can not say anything about further work with this, but at the moment to solve this problem, simply replace the scrollToBottom method with this one



     public func scrollToBottomAnimated(animated: Bool) {
        
        guard numberOfSections > 0 else{
            return
        }
        
        let items = numberOfItems(inSection: 0)
        
        if items == 0 { return }
        
        let collectionViewHeight = collectionViewLayout.collectionViewContentSize.height
        
        scrollRectToVisible(CGRect(x: 0, y: collectionViewHeight - 1, width: 1, height: 1), animated: animated)
    }

**Comment by jackdem** [comment]
did you put this in the framework its self or in the viewcontroller in app?

**Comment by SD10** [comment]
@jackdem he put it in the framework itself

**Comment by jackdem** [comment]
can we get a pull request update? be good to get this issue resolved

**Comment by SD10** [comment]
@jackdem I’m pretty sure that’s not the fix, hopefully it’s fixed already in the next release. You can check the `v0.10.0` example app to see if this issue is still present.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by Terens777** [comment]
As I wrote this is a temporary solution

**Comment by stefanluptak** [comment]
As of the weird `scrollToBottom` behaviour, all what I had to adjust was to change line
`scrollToItem(at: indexPath, at: .bottom, animated: animated)`
to
`scrollToItem(at: indexPath, at: .top, animated: animated)`
inside method `scrollToBottom(animated:)` in file `MessagesCollectionView.swift`
(difference is in `.bottom` vs `.top` (see: https://developer.apple.com/documentation/uikit/uicollectionviewscrollposition)

**Comment by jackdem** [comment]
that lets me scroll to the top which works, but thats not really useful for a conversation? you would want to go to the bottom, which isnt working still

**Comment by stefanluptak** [comment]
@jackdem did you try it? For me it solved the situation when the bottom message is partialy visible and partialy hidden.

**Comment by Terens777** [comment]
in 30 minutes I will return for the workplace and take off the option as I decided this problem

**Comment by jackdem** [comment]
yes i tried, for me it loads the conversation at the top only, opening the keyboard still crops the conversation and not auto scrolling to the latest messages in a messaging app isnt good

**Comment by SD10** [comment]
Please let me know if this is still happening in `v0.10.0`. I'm going to look at `UICollectionViewScrollPosition` too.

**Comment by jackdem** [comment]
im on the latest version on cocapods whichever that is. nothings resolved or improved the situation for me

**Comment by SD10** [comment]
The problem is when `scrollToBottom(animated:)` is called for `isFirstLayout`, the `messagesCollectionView.numberOfSections` is zero. I have to figure out what caused this because it was working before 🤔 I even had to disable it in the example because it was annoying me that the page opened at the bottom 😅

**Comment by Terens777** [comment]
at the moment I use it this way




    public func scrollToBottom(animated: Bool = false) {
        
        guard let indexPath = indexPathForLastItem else { return }
        scrollToItem(at: indexPath, at: .bottom, animated: animated)
    }
    public func scrollToBottomAnimated(animated: Bool) {
        
        guard numberOfSections > 0 else{
            return
        }
        
        let items = numberOfItems(inSection: 0)
        
        if items == 0 { return }
        
        let collectionViewHeight = collectionViewLayout.collectionViewContentSize.height
        
        let isContentTooSmall: Bool = (collectionViewHeight < frame.size.height)
        
        if isContentTooSmall {
            
            scrollRectToVisible(CGRect(x: 0, y: collectionViewHeight - 1, width: 1, height: 1), animated: animated)
            return
        }
        
        scrollToBottom(animated: true)
    }

**Comment by bayareahank** [comment]
Tried the latest release (0.10), when message is just one line, screen scrolls up correctly and show the new message. When longer than one line, screen just won't scroll up, the new message is thus hidden.

**Comment by SD10** [comment]
Hey @Terens777, thanks for providing that code. I'm wondering why you perform the check `isContentTooSmall` and then fall back on `scrollRectToVisible`? 

Could we not just use `scrollRectToVisible` 100% of the time?

**Comment by SD10** [comment]
This is proving pretty difficult to solve. #301 fails to resolve the issue consistently. 

The code snippet provided by @Terens777 doesn't seem to work without being wrapped in `Dispatch.main.async`

**Comment by festrs** [comment]
I found some topic on stack overflow that might help!

https://stackoverflow.com/questions/17751985/animated-scroll-to-item-in-uicollectionview-doesnt-always-work

I'll try to implement a fix and them submit a PR!

**Comment by SD10** [comment]
@festrs #395 Attempts to solve this one, thank you for taking interest 👍

**Comment by SD10** [comment]
This should be resolved in #395 when 0.12.0 is released
